### ISSUE TITLE
Claim-splitting inside original claim, not top-level

### ISSUE THREAD

**Issue by PetarMax** [runtimeverification/haskell-backend#3377]
If an initial simplification of a given claim yields a split (e.g., into two claims), this split is performed at top level, and there will be two separate claims to be proven. This works well if the proof does not require circularity, but otherwise the claim will become unprovable.

I believe that the claim should not be split top-level, but that the splitting should happen under the umbrella of the original claim, and the only circularity applicable should be on the top-level claim.

**Comment by ana-pantilie** [comment]
Duplicate of [#3312](https://github.com/runtimeverification/haskell-backend/issues/3312)

**Comment by PetarMax** [comment]
I'm not sure this is a duplicate of #3312, my suggestion is to definitely do the pre-processing but just not split the top-level claim. Perhaps this is intertwined in a way that I don't fully understand.

**Comment by ana-pantilie** [comment]
@PetarMax sorry for not being clear, pre-processing means simplifying and simplification can split patterns into two. That's the reason you're getting multiple claims out of one.

**Comment by ehildenb** [comment]
But I think the point here is that we _want_ the preprocessing for discharging the claims (we _want_ to do the case split to discharge), but we want to use the _original_ claim as a circularity.

So if we have `(A #Or B) => C`, we _want_ the backend to split and discharge `A => C` and `B => C`, but we want the original claim `(A #Or B) => C` to be used as the circularity for discharging other proof obligations, _not_ the split claims.

**Comment by ehildenb** [comment]
Reopening this because it's more about proof strategy than about processing the claims at startup.

**Comment by ana-pantilie** [comment]
Claims are already being simplified, whether we do this at start-up (in Kore.Exec) or during the first execution step (in Kore.Reachability.Claim) is something we need to figure out. But, I believe that if we don't simplify in Kore.Exec then this and the other issue, will be fixed. This is currently being tried out. If that is problematic, then yes, we will have to keep the original claims around, link them to the processed claims and use the original inside the `ApplyClaims` transition.

That being said, this _is_ a duplicate of the other issue, so I will close this. If you can provide more information that will make it obvious that the two issues are not related, then I will gladly reopen.

**Comment by ehildenb** [comment]
Then we can leave this open until the matter is resolved? It's still not clear to me that the solution to the other issue will be a solution to this one as well. In the other one, Amelie seems to be pointing out a bug in proving, and this one is about proof strategy.

I don't see why we should close this though, as it's not resolved, and it needs to be resolved for the Plutus proofs.

**Comment by ana-pantilie** [comment]
He pointed out this issue a while ago in a Slack thread. I did not open a new issue back then because I knew Amelie's issue was closely related and I bumped the priority of that one immediately.

I didn't close this as resolved but as a duplicate.

> Amelie seems to be pointing out a bug in proving, and this one is about proof strategy

I mean, my reasoning is that both issues come from _losing the original claims, the way the user wrote them_ along the way.

We've done some experimentation (results visible on the other issue) to see the extent of how much modification is done to the original claims.

I will reopen because you think it should be tracked separately.
